### PR TITLE
[1LP][WIP] VM Widgetastic Conversion

### PR DESCRIPTION
**Pull request by rsnyman**
Convert Virtual Machines to Widgetastic

### CODE REVIEW

**Comment by rsnyman** [comment]
The failures appear to be related to bugs in 5.9.

**Comment by psav** [comment]
@rsnyman Really want to get this merged as soon as we can

**Comment by rsnyman** [comment]
Only errors seem to be a timeout waiting for an external server, and an endpoint issue unrelated to my changes. @psav if you're happy, you can merge this.

File: cfme/infrastructure/virtual_machines.py [hunk 1]
```diff
@@ -1080,9 +1085,9 @@ def remove(vm_names, cancel=True, provider_crud=None):
         cancel: Whether to cancel the deletion, defaults to True
         provider_crud: provider object where vm resides on (optional)
     """
-    _method_setup(vm_names, provider_crud)
-    cfg_btn('Remove selected items from the VMDB', invokes_alert=True)
-    sel.handle_alert(cancel=cancel)
+    view = _method_setup(vm_names, provider_crud)
+    view.toolbar.configuration.item_select('Remove selected items from the VMDB',
```
Comment (izapolsk) [inline cfme/infrastructure/virtual_machines.py]: Q: are you sure this is true for both 5.8 and 5.9 ?

File: cfme/infrastructure/virtual_machines.py [hunk 2]
```diff
@@ -991,39 +994,40 @@ def compare(self, *objects, **kwargs):
...
+        view = self.navigate()
+        return view.tree.read_contents()
 
     @property
     def ancestors(self):
         """Returns list of ancestors of the represented object."""
-        self.navigate()
-        path = self.genealogy_tree.find_path_to(re.compile(r"^.*?\(Selected\)$"))
+        view = self.navigate()
+        path = view.tree.find_path_to(re.compile(r"^.*?\(Selected\)$"))
```
Comment (izapolsk) [inline cfme/infrastructure/virtual_machines.py]: REQUIRED: please double check this code. WT tree doesn't have methods like find_path_to

File: cfme/infrastructure/virtual_machines.py [hunk 3]
```diff
@@ -991,39 +994,40 @@ def compare(self, *objects, **kwargs):
...
+        assert len(objects) >= 2, 'You must specify at least two objects'
         objects = map(lambda o: o.name if isinstance(o, (Vm, Template)) else o, objects)
-        self.navigate()
+        view = self.navigate()
         for obj in objects:
             if not isinstance(obj, list):
-                path = self.genealogy_tree.find_path_to(obj)
-            self.genealogy_tree.check_node(*path)
-        toolbar.select("Compare selected VMs")
+                path = view.tree.find_path_to(obj)
```
Comment (izapolsk) [inline cfme/infrastructure/virtual_machines.py]: REQUIRED: please check this code. as far as I see you use WT tree here.

File: cfme/infrastructure/virtual_machines.py [hunk 4]
```diff
@@ -309,6 +251,72 @@ class InfraVmReconfigureView(BaseLoggedInPage):
...
+    name = TextInput('name')
+    description = TextInput('description')
+    snapshot_vm_memory = Checkbox('snap_memory')
+    create = Button('Create')
+    cancel = Button('Cancel')
+
+    @property
+    def is_displayed(self):
+        """Is this view being displayed"""
+        return (
```
[the remark this answers is not in the document]
Comment (psav) [reply cfme/infrastructure/virtual_machines.py]: Yeh I think I agree on that, we have had that issue in the past

File: cfme/infrastructure/virtual_machines.py [hunk 5]
```diff
@@ -955,31 +964,25 @@ class Genealogy(object):
...
+        'different': 'Attributes with different values',
+        'same': 'Attributes with same values',
     }
 
-    def __init__(self, o):
-        self.o = o
+    def __init__(self, obj):
+        self.obj = obj
 
     def navigate(self):
```
[the remark this answers is not in the document]
Comment (psav) [reply cfme/infrastructure/virtual_machines.py]: Yeh this should be removed, but I think it is to maintain backwards compat and I'm OK with that right now

File: cfme/infrastructure/virtual_machines.py [hunk 6]
```diff
@@ -797,25 +808,27 @@ def total_snapshots(self):
...
-            "//li[contains(normalize-space(.), '(Active)')]").strip()
-        # In 5.6 the locator returns the entire tree string, snapshot name is after a newline
-        return re.sub(r"\s*\(Active\)$", "", text.split('\n')[-1:][0])
+        view = navigate_to(self, 'SnapshotsAll')
+        if not view.tree.is_displayed:
+            return None
+        root_item = view.tree.expand_path(self.name)
+        active_items = view.tree.child_items_with_text(root_item, '(Active)')
+        if active_items:
+            return active_items[0].text.split(' (Active')[0]
```
[the remark this answers is not in the document]
Comment (psav) [reply cfme/infrastructure/virtual_machines.py]: What's the typo here?

File: cfme/infrastructure/virtual_machines.py [hunk 7]
```diff
@@ -836,40 +849,36 @@ def cluster_id(self):
...
         def set_relationship(self, server_name, server_id, click_cancel=False):
-            self.navigate()
-            option = "{} ({})".format(server_name, server_id)
+            view = self.navigate()
+            view.form.fill({'server': '{} ({})'.format(server_name, server_id)})
 
             if click_cancel:
-                fill(self.relationship_form, {'server_select': option},
-                     action=self.relationship_form.cancel_button)
+                view.form.cancel_button.click()
```
[the remark this answers is not in the document]
Comment (psav) [reply cfme/infrastructure/virtual_machines.py]: In light of some recent discussions I think we can ignore this for now

File: cfme/infrastructure/virtual_machines.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (psav) [reply cfme/infrastructure/virtual_machines.py]: So we'll leave it in for now

File: cfme/infrastructure/virtual_machines.py [hunk 8]
```diff
@@ -836,40 +849,36 @@ def cluster_id(self):
...
                 ('reset_button', form_buttons.reset),
                 ('cancel_button', form_buttons.cancel)
             ])
 
-        def __init__(self, o):
-            self.o = o
+        def __init__(self, vm):
+            self.vm = vm
 
         def navigate(self):
```
[the remark this answers is not in the document]
Comment (psav) [reply cfme/infrastructure/virtual_machines.py]: Agreed but in here for backwards compat and that's ok for now

File: cfme/infrastructure/virtual_machines.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (izapolsk) [reply cfme/infrastructure/virtual_machines.py]: line above - '(Active)'
current line - '(Active'
This is not typo but rather difference which distracts a bit when look thru code.

**Pull request merged**